contracts/timealarms/src/alarms.rs

```diff
@@ -3,92 +3,108 @@ use platform::{
+const REPLY_ID: Id = 0;
+const EVENT_TYPE: &str = "timealarm";
+
+pub(super) struct TimeAlarms<'r> {
+    time_alarms: Alarms<'r>,
 }
 
-impl TimeAlarms {
-    const ALARMS_NAMESPACE: &'static str = "alarms";
-    const ALARMS_IDX_NAMESPACE: &'static str = "alarms_idx";
-    const REPLY_ID: Id = 0;
-    const EVENT_TYPE: &str = "timealarm";
+pub(super) struct TimeAlarmsMut<'r> {
+    time_alarms: AlarmsMut<'r>,
+}
 
-    pub(super) fn new() -> Self {
+impl<'r> TimeAlarms<'r> {
+    pub fn new(storage: &'r dyn Storage) -> Self {
         Self {
-            time_alarms: Alarms::new(Self::ALARMS_NAMESPACE, Self::ALARMS_IDX_NAMESPACE),
+            time_alarms: Alarms::new(storage, ALARMS_NAMESPACE, ALARMS_IDX_NAMESPACE),
         }
     }
 
-    pub(super) fn remove(
-        &self,
-        storage: &mut dyn Storage,
-        addr: Addr,
-    ) -> Result<(), ContractError> {
-        self.time_alarms.remove(storage, addr)?;
+    pub fn try_any_alarm(&self, ctime: Timestamp) -> Result<AlarmsStatusResponse, ContractError> {
+        let remaining_alarms = self
+            .time_alarms
+            .alarms_selection(ctime)
+            .next()
+            .transpose()?
+            .is_some();
+
+        Ok(AlarmsStatusResponse { remaining_alarms })
+    }
+}
+
+impl<'r> TimeAlarmsMut<'r> {
+    pub fn new(storage: &'r mut dyn Storage) -> Self {
+        Self {
+            time_alarms: AlarmsMut::new(storage, ALARMS_NAMESPACE, ALARMS_IDX_NAMESPACE),
+        }
+    }
+
+    pub fn remove(&mut self, addr: Addr) -> Result<(), ContractError> {
+        self.time_alarms.remove(addr)?;
+
         Ok(())
     }
 
-    pub(super) fn try_add(
-        &self,
-        deps: DepsMut<'_>,
-        env: Env,
-        address: Addr,
+    pub fn try_add(
+        &mut self,
+        querier: &QuerierWrapper<'_>,
+        env: &Env,
+        subscriber: Addr,
         time: Timestamp,
     ) -> ContractResult<MessageResponse> {
         if time < env.block.time {
             return Err(ContractError::InvalidAlarm(time));
         }
-        contract::validate_addr(&deps.querier, &address)
+
+        contract::validate_addr(querier, &subscriber)
             .map_err(ContractError::from)
-            .and_then(|()| {
-                self.time_alarms
-                    .add(deps.storage, address, time)
-                    .map_err(Into::into)
-            })
+            .and_then(|()| self.time_alarms.add(subscriber, time).map_err(Into::into))
             .map(|()| Default::default())
     }
 
-    pub(super) fn try_notify(
-        &self,
-        storage: &mut dyn Storage,
+    pub fn try_notify(
+        &mut self,
         ctime: Timestamp,
         max_count: AlarmsCount,
     ) -> ContractResult<(AlarmsCount, MessageResponse)> {
         self.time_alarms
-            .alarms_selection(storage, ctime)
+            .alarms_selection(ctime)
             .take(max_count.try_into()?)
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
```

Comment:
IMHO we have to bail out with an error if we detect that there are alarms left in-delivery